### DISCUSSION TITLE
Support using the back button

### DISCUSSION THREAD

**Discussion by ju5t** [spatie/laravel-livewire-wizard#27] [Ideas]
This is not something I have time to work on immediately, but it's something we would need ourselves in the future. You could build this with JS; you'd change the url hash when the page loads and immediately after add a listener for `hashchange`. There may be other/better solutions, I'm thinking aloud here :)

I haven't looked into this in great detail but I thought I'd raise it here as a reminder for myself and perhaps someone else who does have time to make this in the future.

**Comment by cbaconnier** [comment]
I have made this trait;

- It have nice URL and doesn't show the full component name: `example.org?step=confirmation`
- Since it uses Livewire QueryString, it support navigation back, forth and refresh

There's a quirk tough: I have to override `showStep()` in order to update the QueryString parameter to use nicer URLs.
But still, I suppose this trait could be added as PR? 

```php
trait WizardQueryString
{
    public string|null $currentStepQueryString = null;
    public string $queryStringStepName = 'step';

    public function queryStringWizardQueryString()
    {
        return [
            'currentStepQueryString' => ['as' => $this->queryStringStepName],
        ];
    }

    public function initializeWizardQueryString()
    {
        // Select the correct step from the URL Query String
        if ($step = collect($this->queryStringSteps())->search(request()->query($this->queryStringStepName))) {
            $this->currentStepName = $step;
        }
    }

    public function showStep($toStepName, array $currentStepState = [])
    {
        parent::showStep($toStepName, $currentStepState);

        // Update the URL Query String
        if ($parameter = $this->queryStringSteps()[$toStepName] ?? null) {
            $this->currentStepQueryString = $parameter;
        }
    }

    protected function queryStringSteps(): array
    {
        // To have "nice" URL parameters
        return [

        ];
    }
}
```

Usage
```php
class RegisterWizard extends WizardComponent
{
    use WizardQueryString;

    public function steps(): array
    {
        return [
            RegisterWinesStep::class,
            RegisterAddressStep::class,
            RegisterConfirmationStep::class,
        ];
    }

    protected function queryStringSteps(): array
    {
        return [
            Livewire::getAlias(RegisterWinesStep::class)        => 'wines',
            Livewire::getAlias(RegisterAddressStep::class)      => 'address',
            Livewire::getAlias(RegisterConfirmationStep::class) => 'confirmation',
        ];
    }
}
```
--- 
A simpler version would be to use `$queryString` on `currentStepName` but would make a bit ugly

`example.org?currentStepName=register-wizard-steps.register-wines-step` :face_with_spiral_eyes: 

```php
class RegisterWizard extends WizardComponent
{
    protected $queryString = [
        'currentStepName'
    ];

    public function steps(): array
    {
        return [
            RegisterWinesStep::class,
            RegisterAddressStep::class,
            RegisterConfirmationStep::class,
        ];
    }
}
```

**Reply by beholdr** [reply]
Looks like this method is not saving state when using Back button. For example:
- enter data on first step, go to the next step, press browser Back button = field on first step is empty
- enter data on first step, go to the next step, return by `previousStep()` = field on first step is filled

![May-06-2023 16-08-46](https://user-images.githubusercontent.com/741973/236626679-3ac78e0b-5e54-4f9a-9a9c-93f892d7e1da.gif)

Livewire v2.12.3 and laravel-livewire-wizard v1.2.0

**Reply by cbaconnier** [reply]
Nice catch! I didn't noticed that. I think I was storing the data when I used it. Refreshing the page would preserve the state too.
I don't work very much with Livewire at the moment. I don't know if we can have a better back button support.
